fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveUnionCursor.java

```diff
@@ -0,0 +1,323 @@
+                return from(parsed, tempTableDeserializer);
+            } catch (InvalidProtocolBufferException ex) {
+                throw new RecordCoreException("invalid continuation", ex)
+                        .addLogInfo(LogMessageKeys.RAW_BYTES, ByteArrayUtil2.loggable(unparsedContinuationBytes));
+            }
+        }
+
+        public boolean isInitialState() {
+            return isInitialState;
+        }
+
+        @Nonnull
+        public RecordCursorContinuation getActiveStateContinuation() {
+            return activeStateContinuation;
+        }
+
+        @Nonnull
+        public TempTable getTempTable() {
+            return tempTable;
+        }
+
+        public boolean buffersAreFlipped() {
+            return buffersAreFlipped;
+        }
+    }
+
+    /**
+     * Interface for recursive state management, the caller is expected to invoke respective callbacks when certain
+     * events occur so the internal state mutates accordingly, it also offers a set of methods that enable examining
+     * current state and whether the execution can proceed to a subsequent recursive state or not.
+     * @param <T> The type of the cursor elements.
+     */
+    public interface RecursiveStateManager<T> {
+
+        /**
+         * Callback notifying the manager that the current cursor being iterated over by the {@link RecursiveUnionQueryPlan}
+         * is exhausted.
+         */
+        void notifyCursorIsExhausted();
+
+        /**
+         * Checks whether it is possible to transition from current recursive step {@code n} to next step {@code n + 1}.
+         * @return {@code True} if it is possible to transition from current recursive step {@code n} to next step {@code n + 1},
+         * otherwise {@code false}.
+         */
+        boolean canTransitionToNewStep();
+
+        /**
+         * Retrieve the currently active cursor, an active cursor is either the {@code initial} cursor or the {@code recursive} cursor.
+         * @return The currently active cursor.
+         */
+        @Nonnull
+        RecordCursor<T> getActiveStateCursor();
+
+        /**
+         * Retrieve the {@link TempTable} that is owned by the {@link RecursiveUnionQueryPlan}.
+         * @return The {@link TempTable} that is owned by the {@link RecursiveUnionQueryPlan}.
+         */
+        @Nonnull
+        TempTable getRecursiveUnionTempTable();
+
+        /**
+         * Checks whether the execution is still in the initial phase or not.
+         * @return {@code True} if the execution is still in the initial phase, otherwise {@code false}, i.e. the execution
+         * is in the recursive phase.
+         */
+        boolean isInitialState();
+
+        /**
+         * Checks whether read and write buffers ({@link TempTable}s) are swapped or not.
+         * @return {@code True} if the read and write buffers are swapped, otherwise {@code false}.
+         */
+        boolean buffersAreFlipped();
```

Comment:
There is something strange about having this as part of the manager contract. Is it that it makes the ownership pf the temp tables simpler / more stable? And avoids the need to change the resolution of quantifiers for the read and insert sets?

It seems like, perhaps with richer snapshot semantics alluded to elsewhere, the operation of replacing the contents of the read table with the contents of the write table and emptying the write table could be done so as not to have to know which buffer was which. If so, then perhaps this can be addressed later.

There does not seem to be any actual problem with the implementation.

Reply:
I removed this method. In a previous attempt to implement the continuation semantics, I wanted to serialize the temp table ownership itself and the role of each temp table. However it turned out that this is not necessary, as the recursive union always owns the read buffer and the insert operators underneath own the write buffer. This invariant is held across continuations, and the actual buffer flip can be thought of as mere runtime routine used to trigger the calculation of the subsequent recursive step (if any).

Reply:
Considering that `buffersAreFlipped` method is removed, and swapping the buffer is something done internally by the state manager, I think we could just create a new temp table instead of cleaning it when flipping the buffers when using a temp-table snapshot-like strategy instead of the current expensive approach:

```java
private EvaluationContext flipBuffers(@Nonnull final EvaluationContext evaluationContext) {
    final var insertTempTable = getTempTable(evaluationContext, insertTempTableReference);
    final var scanTempTable = getTempTable(evaluationContext, scanTempTableReference);
    if (recursiveUnionTempTable == insertTempTable) {
        recursiveUnionTempTable = scanTempTable;
        insertTempTable.clear(); // -- replace with: insertTempTable = new TempTable();
        return overrideTempTableBinding(
                overrideTempTableBinding(baseContext, insertTempTableReference, insertTempTable),
                scanTempTableReference, scanTempTable);
    } else {
        recursiveUnionTempTable = insertTempTable;
        scanTempTable.clear(); // -- replace with: scanTempTable = new TempTable();
        return overrideTempTableBinding(
                overrideTempTableBinding(baseContext, insertTempTableReference, scanTempTable),
                scanTempTableReference, insertTempTable);
    }
}
```